### ISSUE TITLE
Custom XML entities not supported

### ISSUE THREAD

**Issue by trevorrowe** [ohler55/ox#122]
I've run into an issue attempting to use Ox when parsing XML documents. Currently the project uses nokogiri and loads the XML documents into a DOM where I crawl the tree of elements. I was in the process of porting over to Ox when I ran into issues with the documents that use custom xml entities. For example:

```ruby
require 'ox'
require 'stringio'

xml = <<-XML.strip
<?xml version="1.0" standalone="yes" ?>
<!DOCTYPE example [
  <!ENTITY example "example-value">
]>
<xml>
  <works>&example;</works>
  <oops bad="&example;"></oops>
</xml>
XML

class SaxParser

  def attr(name, str); end

  def error(msg, line = nil, column = nil)
    raise "#{msg} on line #{line} at column #{column}"
  end

end

Ox.sax_parse(SaxParser.new, StringIO.new(xml), convert_special: true)
#=> raises RuntimeError: Not Terminated: special character does not end with a semicolon on line 7 at column 18
```

Please note, the custom entity does not raise an issue if it appears in the text of an XML element, but if it appears in the value of an XML attribute, then an error is generated. Ideally I should be able to sax parse the doctype and its nested entity declarations. Also, custom entities should be permitted in attribute values in the same way they can appear in text. Let me know if I need to provide some additional context.

Thanks!

**Comment by ohler55** [comment]
Should have raised an exception in an element as well in the current version. I'll make sure that is consistent. More important I will add the entity registration.

**Comment by trevorrowe** [comment]
Should it raise an error if the entity is not defined in the current document? I have other documents that I've worked with that use an external document for defining entities. Would it be sufficient to simply leave the entities and not raise?

**Comment by ohler55** [comment]
I think it should raise if the entity is not defined. I can add a option to allow entities to be defined as an option hash parameter though. That should handle the case where you want to predefine entities.

**Comment by trevorrowe** [comment]
There are few edge case I don't think providing a hash of entity definitions will solve:

* The doctype contains a URL to en external entity declaration (more information here: https://xmlwriter.net/xml_guide/entity_declaration.shtml)
* The entity definitions are recursive

I run into both of these situations with the XML documents I am working with. Sometimes the entities provide a URL to an external document that I download and extract entity definitions from. These can then be used to replace entities found in the XML text and attribute values.

The second happens when one entity definition contains another custom entity. These may reference sometimes 4-5 times deep in the documents I work with.

Would it be feasible to expose the entity definitions as XML parsing events (or as children in the document) and then leave the custom entities intact?

**Comment by ohler55** [comment]
There is an option to the SAX parser called convert_special. If it is false then &xxx; sequences are not converted nor are there errors generated. I noticed that it dis not apply to attributes so I have fixed that in the github version. I can also make that option effective with the general parser.

When that option is working then all that is left is the handling of the DOCTYPE element. Right now any parsing of the DOCTYPE contents is left up to the caller. If that works for you this should be a pretty quick fix. If not it will take a little longer to parse DOCTYPE contents.
